### ISSUE TITLE
Add Semaphore, Condition, Event

### ISSUE THREAD

**Issue by mrocklin** [dask/distributed#2007]
Dask currently has some distributed coordination primitives like Locks and Queues that match the API of normal local Locks and Queues.  There are other primitives like these that we do not yet support, notable `Semaphore`, `Condition`, and `Event`.

These might be good training for people who want to become familiar with extending the dask.distributed system.  The files `lock.py`, `queues.py`, and `variable.py` might be a good model for development here.

**Comment by palash25** [comment]
Hi, I've been writing python for some time but I'm new to distributed computing. This seems like the right issue to get started. Can I work on this?

**Comment by mrocklin** [comment]
Yes!  Anyone can work on any issue.  You may want to familiarize yourself
with http://distributed.readthedocs.io/en/latest/develop.html

**Comment by palash25** [comment]
Thanks will try to submit a PR within a week's time

**Comment by andersy005** [comment]
- It appears that https://github.com/dask/distributed/pull/2690 is/was meant to implement one (Semaphore) of the missing features listed here. Are there any updates on the status of https://github.com/dask/distributed/pull/2690? 

- https://github.com/dask/distributed/pull/3397 mentions Events and Conditions. Is it the case that these two features have also been implemented already?

I am interested in working on the other features namely (Event, Condition), and wanted to make sure that I am not duplicating work that has already been done.

**Comment by mrocklin** [comment]
> It appears that #2690 is/was meant to implement one (Semaphore) of the missing features listed here. Are there any updates on the status of #2690?

All communication is happening on github.  I recommend checking in on that PR.

> #3397 mentions Events and Conditions. Is it the case that these two features have also been implemented already?

This is different.  Dask uses Tornado/asyncio objects within its codebase.  We would also like to expose distributed versions of these objects to users so that they can make their own applications.

**Comment by andersy005** [comment]
> All communication is happening on github. I recommend checking in on that PR.

👍  will do 

> This is different. Dask uses Tornado/asyncio objects within its codebase. We would also like to expose distributed versions of these objects to users so that they can make their own applications.

Thank you for the clarification!

**Comment by nadzhou** [comment]
Can i work on the `Condition` part, it seems like `Event` has been implemented.
